pdf-service/app/render_cache.py

```
@@ -0,0 +1,173 @@
+def _key_path(input_bytes: bytes, ext: str) -> Path:
+    """sha256(input) + ext → cache file path。"""
+    assert _cache_dir is not None
+    digest = hashlib.sha256(input_bytes).hexdigest()
+    safe_ext = ext.lstrip(".").lower()
+    if safe_ext not in {"docx", "xlsx", "pdf"}:
+        safe_ext = "bin"
+    return _cache_dir / f"{digest}.{safe_ext}.pdf"
+
+
+def get(input_bytes: bytes, ext: str) -> bytes | None:
+    """Cache lookup。命中 → 回 PDF bytes；未命中或過期 → None。
```

Ruff (0.15.12)

[warning] Docstring contains ambiguous `；` (FULLWIDTH SEMICOLON). Did you mean `;` (SEMICOLON)?

(RUF002)